internal/appsec/remoteconfig.go

```diff
@@ -83,6 +83,65 @@ func (a *appsec) asmFeaturesCallback(u remoteconfig.ProductUpdate) map[string]rc
 	return statuses
 }
 
+type wafHandleWrapper struct {
+	*waf.Handle
+}
+
+func (h *wafHandleWrapper) asmDataCallback(u remoteconfig.ProductUpdate) map[string]rc.ApplyStatus {
+	// Following the RFC, merging should only happen when two rules data with the same ID and same Type are received
+	// rulesData[ID][Type] will return the rules data of said id and type, if it exists
+	allRulesData := make(map[string]map[string]rc.ASMDataRuleData)
+	statuses := statusesFromUpdate(u, true, nil)
+
+	for path, raw := range u {
+		log.Debug("appsec: Remote config: processing %s", path)
+		var rulesData rc.ASMDataRulesData
+		if err := json.Unmarshal(raw, &rulesData); err != nil {
+			log.Debug("appsec: Remote config: error while unmarshalling payload for %s: %v. Configuration won't be applied.", path, err)
+			statuses[path] = genApplyStatus(false, err)
+			continue
+		}
+
+		// Check each entry against allRulesData to see if merging is necessary
+		for _, ruleData := range rulesData.RulesData {
+			if allRulesData[ruleData.ID] == nil {
+				allRulesData[ruleData.ID] = make(map[string]rc.ASMDataRuleData)
+			}
+			if data, ok := allRulesData[ruleData.ID][ruleData.Type]; ok {
+				allRulesData[ruleData.ID][ruleData.Type] = mergeRulesData(ruleData, data)
+			} else {
+				allRulesData[ruleData.ID][ruleData.Type] = ruleData
+			}
+		}
+	}
+
+	// Aggregate all the rules data before passing it over to the WAF
+	var rulesData []rc.ASMDataRuleData
+	for _, m := range allRulesData {
+		for _, data := range m {
+			rulesData = append(rulesData, data)
+		}
+	}
+
+	payload, err := json.Marshal(rulesData)
+	if err != nil {
+		log.Debug("appsec: Remote config: could not marshal the merged rules data: %v.", err)
+		statuses = statusesFromUpdate(u, false, err)
+	} else if err := h.UpdateRuleData(payload); err != nil {
+		log.Debug("appsec: Remote config: could not update WAF rule data: %v.", err)
+		statuses = statusesFromUpdate(u, false, err)
+	}
```

Comment:
This WAF method is unmarshaling the json payload again to convert it into ddwaf_objects.
So we should pass the Go value instead, and adapt the WAF `encoder` so we properly handle it: so far it handles Go structs but we ignore the `json` field tag and we instead use the field name only. I think it would work if we add the json-tag support and use that name instead when it is available.

Reply:
See d6809e6f4b9276